File: Adam_VCF_Join/src/main/scala/VCF_Join_Job.scala
package main.scala

import org.apache.hadoop.io.LongWritable
import org.apache.parquet.hadoop.util.ContextUtil
import org.apache.spark.rdd.RDD
import org.apache.spark.{SparkConf, SparkContext}
import org.bdgenomics.adam.converters.VariantContextConverter
import org.bdgenomics.adam.models.VariantContext
import org.bdgenomics.adam.rdd.ADAMContext
import org.bdgenomics.adam.rdd.read.AlignmentRecordRDD
import org.bdgenomics.formats.avro.AlignmentRecord
import org.bdgenomics.utils.misc.HadoopUtil
import org.seqdoop.hadoop_bam.{VCFInputFormat, VariantContextWritable}

/**
  * master-thesis Clemens Banas
  * Organization: DBIS - University of Innsbruck
  * Created 13.02.16.
  */
object VCF_Join_Job {

  def main(args: Array[String]): Unit = {
    if (args.length != 3) {
      println("usage: spark-submit Adam_FastQ_BaseQual_Baseline-1.0-SNAPSHOT.jar <sample input file> <reference input file> <output dir>")
      return;
    }

    val sample = args(0)
    val reference = args(1)
    val output = args(2)

    val conf = new SparkConf()
    conf.setAppName("Adam_VCF_Join")
    val sc = new SparkContext(conf)


    val ac = new ADAMContext(sc)
    val abc: AlignmentRecordRDD = ac.loadAlignments("filename");

//    abc.recordGroups.


    val job = HadoopUtil.newJob(sc)
    val vcc = new VariantContextConverter()

    //load vcf sample input data into a RDD
    val vcfSampleInput = sc.newAPIHadoopFile(
      sample,
      classOf[VCFInputFormat],
      classOf[LongWritable],
      classOf[VariantContextWritable],
      ContextUtil.getConfiguration(job)
    )

    //convert data into ADAMs VariantContext format
    val sampleRDD: RDD[VariantContext] = vcfSampleInput.flatMap(p => vcc.convert(p._2.get))

    //define the join key on the sample RDD
    val sampleJoinRDD: RDD[Pair[Pair[Int, Int], VariantContext]] = sampleRDD.keyBy( record => VCF_Join_Mapper.mapKeyBy(record.variant) )

    //load vcf reference input data into a RDD
    val vcfReferenceInput = sc.newAPIHadoopFile(
      reference,
      classOf[VCFInputFormat],
      classOf[LongWritable],
      classOf[VariantContextWritable],
      ContextUtil.getConfiguration(job)
    )

    //convert data into ADAMs VariantContext format
    val referenceRDD: RDD[VariantContext] = vcfReferenceInput.flatMap(p => vcc.convert(p._2.get()))

    //define the join key on the sample RDD
    val referenceJoinRDD: RDD[Pair[Pair[Int, Int], VariantContext]] = referenceRDD.keyBy( record => VCF_Join_Mapper.mapKeyBy(record.variant) )

    // join the two RDDs according to their key (composed of chromosome & position)
    val joinedRDD: RDD[Pair[Pair[Int, Int], Pair[VariantContext, Option[VariantContext]]]] = sampleJoinRDD.leftOuterJoin(referenceJoinRDD)

    joinedRDD.map(record => record._1).saveAsTextFile(output)

    // TODO implement result generation
//    val res = joinedRDD.sortBy( record => record._1 ).map( record => VCF_Join_Mapper.constructResult(record._1, record._2) )
//    res.saveAsTextFile(output)
  }

}

File: MapReduce_VCF_Join/src/main/java/VCF_Join_Job.java
import org.apache.hadoop.conf.Configuration;
import org.apache.hadoop.fs.Path;
import org.apache.hadoop.io.NullWritable;
import org.apache.hadoop.mapreduce.Job;
import org.apache.hadoop.mapreduce.lib.input.FileInputFormat;
import org.apache.hadoop.mapreduce.lib.output.FileOutputFormat;
import org.seqdoop.hadoop_bam.VCFInputFormat;
import util.*;

/**
 * master-thesis Clemens Banas
 * Organization: DBIS - University of Innsbruck
 * Created 21.04.2016
 */
public class VCF_Join_Job {

    public static void main(String[] args) throws Exception {
        if (args.length != 3) {
            System.out.println("usage: hadoop jar Mapreduce_VCF_Join-1.0-SNAPSHOT.jar <sample vcf file> <reference vcf file> <output dir>");
            return;
        }

        Configuration conf = new Configuration();
        conf.set("sample", args[0].substring(args[0].lastIndexOf('/')+1));
        conf.set("reference", args[1].substring(args[1].lastIndexOf('/')+1));
        Job job = Job.getInstance(conf, "MR_VCF_ReduceSideJoin");

        job.setJarByClass(VCF_Join_Job.class);
        job.setMapperClass(VCF_Join_Mapper.class);
        job.setReducerClass(VCF_Join_Reducer.class);

        //partitioning, grouping and secondary sort classes
        job.setPartitionerClass(KeyPartitioner.class);
        job.setGroupingComparatorClass(NaturalKeyGroupingComparator.class);
        job.setSortComparatorClass(CompositeKeyComparator.class);

        job.setInputFormatClass(VCFInputFormat.class);
        job.setMapOutputKeyClass(ChromPosKey.class);
        job.setMapOutputValueClass(JoinedResultWritable.class);

        job.setOutputKeyClass(NullWritable.class);
        job.setOutputValueClass(JoinedResultWritable.class);

        FileInputFormat.addInputPath(job, new Path(args[0]));
        FileInputFormat.addInputPath(job, new Path(args[1]));
        FileOutputFormat.setOutputPath(job, new Path(args[2]));

        if (!job.waitForCompletion(true)) {
            System.err.println("sort :: Job failed.");
        }
    }

}